ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/jca/XATerminatorImple.java

```diff
@@ -319,12 +324,20 @@ public int prepare (Xid xid) throws XAException
         case XAResource.TMSTARTRSCAN: // check the object store
             if (_recoveryStarted)
                 throw new XAException(XAException.XAER_PROTO);
-            else
+            else {
                 _recoveryStarted = true;
+                if (XARecoveryModule.getRegisteredXARecoveryModule() != null) {
+                    XARecoveryModule.getRegisteredXARecoveryModule().periodicWorkFirstPass();
```

Comment:
If periodicWorkFirstPass is very slow it is possible that the Xids would need refreshing though so I favoured correctness here over optimizing the recovery process